src/lib/spinel/radio_spinel.hpp

```diff
@@ -924,6 +924,10 @@ template <typename InterfaceType> class RadioSpinel
     otError SetChannelTargetPower(uint8_t aChannel, int16_t aTargetPower);
 #endif
 
+#if OPENTHREAD_SPINEL_CONFIG_RCP_RESTORATION_MAX_COUNT > 0
+    void RestoreProperties(void);
```

Comment:
Have you tested this case in your scenario ?

Based on your description, users will directly call the `RestoreProperties()` after the FW if updated when the Thread stack is working.  After the FW is updated, the first Spinel frame sent from the RCP to the host is the RESET reason. The host receives this Spinel frame, it knows that the RCP just got reset, then the host will call  `RecoverFromRcpFailure ()` to restore the RCP. So I think it may not need the user to manually call the function `RestoreProperties ()` to restore the RCP.

Reply:
Yes, I observed this case of `call RecoverFromRcpFailure`, but it is usually a behavior caused by `Unexpected RCP reset`. On the master device side, it will appear as an error. However,this restart behavior is actively caused by the user, so the user may not want `Unexpected RCP reset` to occur. 
In my opinion, if `RecoverFromRcpFailure` can be triggered in some way without showing error, `call RestoreProperties` is not necessary.

Reply:
Even if you manually restore RCP, the Thread host will still can detect the RCP reset and call `RecoverFromRcpFailure `. I think your mainly concern is the error message. You may need to consider how to optimize the log messages.

Reply:
In fact, I must deinitialize the spinel UART between RCP and master device because we use this UART for upadting rcp.

I tried some solutions for updating RCP：

## Solution A
1. Deinitialize the spinel UART.
2. Update the RCP and reset it via RESET pin
3. Initialize the spinel UART.
With solution A, the `RecoverFromRcpFailure` will be triggered.
But this solution will show some errors and warnings:
```
E (15678) OPENTHREAD: dropping radio frame: NoBufs
E(15858) OPENTHREAD:[C] Platform------: Unexpected RCP reset: RESET_POWER_ON
W(15858) OPENTHREAD:[W] Platform------: RCP failure detected
W(15858) OPENTHREAD:[W] Platform------: Trying to recover (1/3)
I(15878) OPENTHREAD:[N] Platform------: RCP recovery is done
```

## Solution B
1. Make the RadioSpinel sleep and disable it vy calling `RadioSpinel<InterfaceType>::Sleep(void)` and `RadioSpinel<InterfaceType>::Disable(void)`.
2. Deinitialize the spinel UART.
3. Update the RCP and reset it via RESET pin.
4. Initialize the spinel UART.
5. Enable the RadioSpinel by calling `RadioSpinel<InterfaceType>::Enable(otInstance *aInstance)`.
With solution A, the `RecoverFromRcpFailure` will also be triggered.
But this solution will show some errors and warnings:
```
E (105748) OPENTHREAD: dropping radio frame: NoBufs
W(107718) OPENTHREAD:[W] Platform------: Wait for response timeout
W(107718) OPENTHREAD:[W] Platform------: RCP failure detected
W(107718) OPENTHREAD:[W] Platform------: Trying to recover (1/3)
I(107738) OPENTHREAD:[N] Platform------: RCP recovery is done
```

## Solution C(unsuitable)
1. Make the RadioSpinel sleep and disable it vy calling `RadioSpinel<InterfaceType>::Sleep(void)` and `RadioSpinel<InterfaceType>::Disable(void)`.
2. Deinitialize the spinel UART.
3. Update the RCP and reset it via RESET pin. **And then delay for a while to wait RCP(such as 1s).**
4. Initialize the spinel UART.
5. Enable the RadioSpinel by calling `RadioSpinel<InterfaceType>::Enable(otInstance *aInstance)`.
With solution A, the `RecoverFromRcpFailure` won't be triggered. So the RCP cannot be recovered, so I tried Solution D


## Solution D
1. Make the RadioSpinel sleep and disable it vy calling `RadioSpinel<InterfaceType>::Sleep(void)` and `RadioSpinel<InterfaceType>::Disable(void)`.
2. Deinitialize the spinel UART.
3. Update the RCP and reset it via RESET pin. **And then delay for a while to wait RCP(such as 1s).**
4. Initialize the spinel UART.
5. Enable the RadioSpinel by calling `RadioSpinel<InterfaceType>::Enable(otInstance *aInstance)`.
6. Restore the RCP by calling `RadioSpinel<InterfaceType>::RestoreProperties(void)`. **(need to expose RestoreProperties as a public method within RadioSpinel)**
7. Set receive if the state of device is not `OT_DEVICE_ROLE_DISABLED` by calling `RadioSpinel<InterfaceType>::Receive(uint8_t aChannel)`.

Solution A、B and D are all suitable, but A and B both cause some errors or warnings. In my opinion, the reset after updating rcp  should not cause these errors or warnings. Shall we tackle the underlying issue rather than merely optimizing the log messages? Or, can some public methods be added to the `RadioSpinel` to allow users to actively restore rcp, such as `RecoverRcpManually`?

Reply:
Concerning RCP restarts, does the Spinel specification include any provisions? Are all RCP restarts classified as `Unexpected RCP reset`?

Reply:
<1> For the UART interface, the RESET reason frame may be missed because the spinel interface has not been initialized. But for the SPI interface, the RESET reason frame won't be missed because the SPI transaction is controlled by the host. We should also consider the SPI interface.

<2> After you deinitialize the spinel interface, the SpinelThread stack sends frames to the RCP.

<3> The initialization of the spinel interface is controlled by [RadioSPinel](https://github.com/zhanglongxia/openthread/blob/master/src/lib/spinel/radio_spinel.cpp#L113). How do you initialize the spinel interface?

Reply:
For <2> and <3>:
When instantiating `RadioSpinel`, `mSpinelInterface` will be assigned. In our program, we point it to an object `UartSpinelInterface` that we implement ourselves.
The `initialize` and `deinitialize` the `spinel interface` I mentioned above do not refer to calling the `Init` and `Deinit` provided by RadioSpinel, but the `Init` and `Deinit` provided by the `UartSpinelInterface` class. These two functions are implement by ourselves and have nothing to do with OpenThread. You can simply understand their function as "acquire/release the current UART resource for spinel interface".
Here is a codes example I am using:
```
// just a example for introduction, does not represent actual code
// s_radio is a RadioSpinel object
s_radio.Sleep();
s_radio.Disable();
s_radio.GetSpinelInterface().Deinit(); // release the current UART resource
rcp_update();
s_radio.GetSpinelInterface().Init(); // acquire the UART resource
s_radio.Enable(instance);
s_radio.Receive();
```

For <1>:
For RadioSpinel, should UART and SPI be transparent? I think when RadioSpinel has released UART/SPI resources, how to avoid receiving error messages when acquire UART/SPI resources again is a concern for platform developers.

Reply:
Thanks for your detailed explanation. The whole process became clearer to me. There are still some corner cases need to be considered:
(1) When the `mState` is `kStateTransmitting` in `radio_spinel.cpp`, calling `s_radio.Sleep()` returns error.  We should consider how to process this case, otherwise, it causes the MAC layer is unable to receive the `otPlatRadioTxDone()` event.
(2) When the `RasioSpinel` is doing the energy scan, updating the FW breaks the scanning process. The function `RestoreProperties ()` won't restore the scanning process.

---

 <1> For RadioSpinel,  UART and SPI should be transparent.  But the working mechanism causes the first RESET frame won't be missed by the SPI interface. One possible solution is to let the developer's code to handle the first RESET frame after the FW is updated, then call the `RestoreProperties ()` to restore the RCP.

Reply:
(1) We will ensure that users are allowed to update rcp only when s_radio.Sleep() returns success.
(2) For the same reason, we will ensure that users are not allowed to update rcp when RasioSpinel is is doing the energy scan.

Reply:
Could you add comments to the function `RestoreProperties ()`?

Reply:
Ok, I have added some comments for it.

Reply:
Regarding the FW update use case. After passing the serial bus back to OpenThread, is it possible to trigger a reset on the RCP again?

It's also not clear to me in what situations the RCP will be in the sleep state. For example, in what cases would the RCP on a border router be in the sleep state?

Reply:
1. In our program design, after the RCP update is completed, the restart of RCP will be completed by some functions other than OpenThread. During this process, any messages generated by RCP should not affect the Host device. After passing the serial bus back to OpenThread, RCP will no longer reboot unless something goes wrong with it itself.
 
2. In fact, what we need to do is `put the radio to sleep and disable it before RCP update`. Therefore, we do not care when the RCP is in the sleep state during the normal operation of the BR.This is my understanding and answer. Did I understand you wrong?

Reply:
If OpenThread stack is running, it assumes that the RCP is responsive. Can you provide more details on how you handle OpenThread being active while the RCP is down and busy doing other things?

Reply:
Thanks for pointing out this situation. I conducted some tests and found that it does cause some problems: if OpenThread tries to send data through spinel when doing RCP update, some errors may be triggered in submac.cpp and cause the program to crash.
To solve this problem, we will add the following logic: before updating rcp, stop thread and interface by calling `otThreadSetEnabled(instance, false)` and `otIp6SetEnabled(instance, false)`.
It is worth mentioning that `RestoreProperties` is still needed to restore RCP information without reinitializing Thread. So I think this pull request is still necessary.

Reply:
@jwhui Are there any other problems about this pull request? Thanks.

Reply:
@gytxxsy if you disable thread and bring interface down (which make sense to do during firmware update), do we still need to have RCP restored by `RadioSpinel`. I expect upon restart of Thread and interface, it will configure radio again (calling all `otPlatRadio{}` APIs).

Reply:
In fact, I did some tests and found the following: When the Host program is initialized, the Host will set the parameters of RCP via spinel; and when starting the interface and Thread, the Host will no longer actively restore RCP. Therefore, without restarting the Host, we need to restore RCP by RadioSpinel.

Reply:
I don't think we need the `RadioSpinel` to restore RCP in such a case.

When we stop Thread operation (`otThreadSetEnabled()`), radio is not used anymore. After start of Thread operation again, radio would be configured again by OT stack (e.g. channel will be set, short address, ect will be set, etc.)

Reply:
Do you mean that the properties of RCP will be restored automaticlly when starting the Interface and Thread? What function is used for this operation? I did not find related operations in the implementation of `otIp6SetEnabled` and `otThreadSetEnabled`. Can you give me a reference link? Thanks.

Reply:
It is the job of the OT stack to configure the radio using the `otPlatRadio` APIs for its use. You can check all places where the `otPlatRadio` APIs  are called (which happen all over the OT stack). 
- To give you one example, `Mle::Start()` calls `Mle::SetRloc16()` then calls `Mac::SetShortAddress()` which end up setting it on radio `otPlatRadioSetShortAddress()`.

`RadioSpinel` is just one example implementation of the `otPlatRadio` API. The RCP recovery mechnsim is intended for when RCP is reset unexpectedt and we don't want the OT stack to be notified and/or restarted, so `RadioSpinel` itself will try to to restore everything so things look same as before.

Reply:
I understand what you mean but when starting Interface and Thread, not all `otPlatRadioxxx()` will be called. 
For example, `otPlatRadioSetPanId()` and `otPlatRadioSetMacKey()` will not be called when starting Interface and Thread. They are only called during program initialization, such as in `otInstanceInitSingle`. So I still think that restoring RCP via `otIp6SetEnable` and `otThreadSetEnable` is not enough. Is there something wrong with my understanding?

Reply:
I see. Thanks. 

I think longer term it may be better to update the OT stack itself to ensure everything is re-initialized by OT stack on radio on Thread restart (this way it will work for `otPlatRadio` API implementation including `RadioSpinel).

Reply:
Your opinion is reasonable, but for users, the resource costs of reinitializing OT stack may be relatively high. Therefore, for application scenarios such as "only update RCP", restoring RCP may be more cost-effective than reinitializing the entire OT stack.

Reply:
@jwhui The change mainly aims to facilitate a customer's use case, allowing for RCP updates without triggering a host SoC reboot.

The general workflow is as follows:
1. The host SoC receives a new RCP firmware and needs to update the RCP without rebooting its own application.
2. The host application stops the Thread by calling `otThreadSetEnabled(instance, false)` and `otIp6SetEnabled(instance, false)`.
3. The RCP update is performed via UART.
4. The host application restarts the Thread stack and calls `RestoreProperties()` to restore the configurations to the RCP.
5. The host application can then continue to work.

Could you please review this workflow to see if there are any concerns?

Reply:
SGTM